extmod/modopenamp.c

```diff
@@ -0,0 +1,394 @@
+    // Flush resource table.
+    metal_cache_flush((uint32_t *)rsc_table, sizeof(openamp_rsc_table_t));
+    #endif
+    *rsc_table_out = rsc_table;
+}
+#endif // MICROPY_PY_OPENAMP_RSC_TABLE_ENABLE
+
+STATIC mp_obj_t openamp_init(uint n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
+    enum { ARG_ns_callback };
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_ns_callback, MP_ARG_OBJ,  {.u_rom_obj = MP_ROM_NONE } },
+    };
+
+    // Parse args.
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all(n_args, pos_args, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+
+    if (MP_STATE_PORT(virtio_device) != NULL) {
+        MP_STATE_PORT(virtio_device)->ns_callback = args[ARG_ns_callback].u_obj;
+        return mp_const_none;
+    }
+
+    struct metal_device *device;
+    struct metal_init_params metal_params = METAL_INIT_DEFAULTS;
+
+    // Initialize libmetal.
+    metal_init(&metal_params);
+
+    // Initialize the shared resource table.
+    openamp_rsc_table_t *rsc_table;
+    openamp_rsc_table_init(&rsc_table);
+
+    if (metal_register_generic_device(&shm_device) != 0) {
+        mp_raise_msg(&mp_type_OSError, MP_ERROR_TEXT("Failed to register metal device"));
+    }
+
+    if (metal_device_open("generic", METAL_SHM_NAME, &device) != 0) {
+        mp_raise_msg(&mp_type_OSError, MP_ERROR_TEXT("Failed to open metal device"));
+    }
+
+    // Initialize shared memory IO region.
+    metal_io_init(&device->regions[0], (void *)METAL_SHM_ADDR, (void *)shm_physmap, METAL_SHM_SIZE, -1U, 0, NULL);
+    struct metal_io_region *shm_io = metal_device_io_region(device, 0);
+    if (shm_io == NULL) {
+        mp_raise_msg(&mp_type_OSError, MP_ERROR_TEXT("Failed to initialize device io region"));
+    }
+
+    // Initialize resource table IO region.
+    metal_io_init(&device->regions[1], (void *)rsc_table, (void *)rsc_table, sizeof(*rsc_table), -1U, 0, NULL);
+    struct metal_io_region *rsc_io = metal_device_io_region(device, 1);
+    if (rsc_io == NULL) {
+        mp_raise_msg(&mp_type_OSError, MP_ERROR_TEXT("Failed to initialize device io region"));
+    }
+
+    // Create virtio device.
+    struct virtio_device *vdev = rproc_virtio_create_vdev(RPMSG_HOST, VIRTIO_DEV_ID,
+        &rsc_table->vdev, rsc_io, NULL, metal_rproc_notify, NULL);
+    if (vdev == NULL) {
+        mp_raise_msg(&mp_type_OSError, MP_ERROR_TEXT("Failed to create virtio device"));
+    }
+
+    // Initialize vrings.
+    struct fw_rsc_vdev_vring *vring_rsc = &rsc_table->vring0;
+    for (int i = 0; i < VRING_COUNT; i++, vring_rsc++) {
+        if (rproc_virtio_init_vring(vdev, vring_rsc->notifyid, vring_rsc->notifyid,
+            (void *)vring_rsc->da, shm_io, vring_rsc->num, vring_rsc->align) != 0) {
+            mp_raise_msg(&mp_type_OSError, MP_ERROR_TEXT("Failed to initialize vrings"));
+        }
+    }
+
+    virtio_dev_obj_t *self = mp_obj_malloc_with_finaliser(virtio_dev_obj_t, &virtio_dev_type);
```

Comment:
It's a bit unconventional to call this `self` which would normally be used for a method on a class. (I was confused reading this because it made me think that I was inside an RPMsg function).

I think it would be clearer if it just used `MP_STATE_PORT(virtio_device)` everywhere in place of `self`.